File: project/jimmer-sql/src/test/java/org/babyfish/jimmer/sql/query/JoinTypeTest.java
package org.babyfish.jimmer.sql.query;

import org.babyfish.jimmer.sql.JoinType;
import org.babyfish.jimmer.sql.ast.query.ConfigurableRootQuery;
import org.babyfish.jimmer.sql.common.AbstractQueryTest;
import org.babyfish.jimmer.sql.model.Book;
import org.babyfish.jimmer.sql.model.BookTable;
import org.junit.jupiter.api.Test;

public class JoinTypeTest extends AbstractQueryTest {

    @Test
    public void testLeftJoin() {
        executeAndExpect(
                createQuery(null),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, tb_1_.NAME, tb_1_.EDITION, tb_1_.PRICE, tb_1_.STORE_ID " +
                                    "from BOOK tb_1_ " +
                                    "left join BOOK_STORE tb_2_ on tb_1_.STORE_ID = tb_2_.ID " +
                                    "order by tb_2_.NAME asc"
                    );
                }
        );
    }

    @Test
    public void testInnerJoin() {
        executeAndExpect(
                createQuery("MANNING"),
                ctx -> {
                    ctx.sql(
                            "select tb_1_.ID, tb_1_.NAME, tb_1_.EDITION, tb_1_.PRICE, tb_1_.STORE_ID " +
                                    "from BOOK tb_1_ " +
                                    "inner join BOOK_STORE tb_2_ on tb_1_.STORE_ID = tb_2_.ID " +
                                    "where tb_2_.NAME ilike ? " +
                                    "order by tb_2_.NAME asc"
                    );
                }
        );
    }

    private ConfigurableRootQuery<BookTable, Book> createQuery(String storeName) {
        BookTable table = BookTable.$;
        return getSqlClient()
                .createQuery(table)
                .where(table.store().name().ilikeIf(storeName))
                .orderBy(table.store(JoinType.LEFT).name().asc())
                .select(table);
    }
}

File: project/jimmer-sql-kotlin/src/test/kotlin/org/babyfish/jimmer/sql/kt/query/JoinTypeTest.kt
package org.babyfish.jimmer.sql.kt.query

import org.babyfish.jimmer.sql.kt.ast.expression.asc
import org.babyfish.jimmer.sql.kt.ast.expression.`ilike?`
import org.babyfish.jimmer.sql.kt.ast.query.KConfigurableRootQuery
import org.babyfish.jimmer.sql.kt.ast.table.KNonNullTable
import org.babyfish.jimmer.sql.kt.common.AbstractQueryTest
import org.babyfish.jimmer.sql.kt.model.classic.book.Book
import org.babyfish.jimmer.sql.kt.model.classic.book.store
import org.babyfish.jimmer.sql.kt.model.classic.book.`store?`
import org.babyfish.jimmer.sql.kt.model.classic.store.name
import org.junit.Test

class JoinTypeTest : AbstractQueryTest() {

    @Test
    fun testOuterJoin() {
        executeAndExpect(
            createQuery(null)
        ) {
            sql(
                """select tb_1_.ID, tb_1_.NAME, tb_1_.EDITION, tb_1_.PRICE, tb_1_.STORE_ID 
                    |from BOOK tb_1_ 
                    |left join BOOK_STORE tb_2_ on tb_1_.STORE_ID = tb_2_.ID 
                    |order by tb_2_.NAME asc""".trimMargin()
            )
        }
    }

    @Test
    fun testInnerJoin() {
        executeAndExpect(
            createQuery("MANNING")
        ) {
            sql(
                """select tb_1_.ID, tb_1_.NAME, tb_1_.EDITION, tb_1_.PRICE, tb_1_.STORE_ID 
                    |from BOOK tb_1_ 
                    |inner join BOOK_STORE tb_2_ on tb_1_.STORE_ID = tb_2_.ID 
                    |where lower(tb_2_.NAME) like ? 
                    |order by tb_2_.NAME asc""".trimMargin()
            )
        }
    }

    private fun createQuery(storeName: String?): KConfigurableRootQuery<KNonNullTable<Book>, Book> =
        sqlClient
            .createQuery(Book::class) {
                where(table.store.name `ilike?` storeName)
                orderBy(table.`store?`.name.asc())
                select(table)
            }
}